File: src_py/tizenclaw/core/ota_updater.py
"""
TizenClaw OTA Updater — Over-the-Air skill update system.

Matches C++ OtaUpdater functionality:
  - Version comparison (semver-style)
  - Remote manifest parsing (check for available updates)
  - Skill download, verify (SHA-256), install with backup
  - Rollback to previous version
  - Auto-check on configurable interval
"""
import asyncio
import hashlib
import json
import logging
import os
import shutil
import ssl
import tarfile
import tempfile
import time
import urllib.request
from typing import Dict, List, Any, Optional
from dataclasses import dataclass, field

logger = logging.getLogger(__name__)

SKILLS_DIR = "/opt/usr/share/tizenclaw/tools/skills"
OTA_CONFIG_PATH = "/opt/usr/share/tizenclaw/config/ota_config.json"
BACKUP_DIR = "/opt/usr/share/tizenclaw/work/ota_backups"


@dataclass
class SkillUpdateInfo:
    name: str = ""
    local_version: str = "0.0.0"
    remote_version: str = "0.0.0"
    url: str = ""
    sha256: str = ""
    update_available: bool = False
    description: str = ""


class OtaUpdater:
    """Over-the-Air skill update manager."""

    def __init__(self, skills_dir: str = SKILLS_DIR,
                 reload_callback=None):
        self._skills_dir = skills_dir
        self._reload_cb = reload_callback
        self._manifest_url = ""
        self._auto_check_hours = 24
        self._auto_update = False
        self._ssl_ctx: Optional[ssl.SSLContext] = None
        self._running = False
        self._task: Optional[asyncio.Task] = None

        try:
            self._ssl_ctx = ssl.create_default_context()
        except Exception:
            self._ssl_ctx = ssl._create_unverified_context()

    # ── Version comparison ──

    @staticmethod
    def is_newer_version(current: str, remote: str) -> bool:
        """Compare semver-style version strings."""
        def parse(v):
            parts = []
            for p in v.split("."):
                try:
                    parts.append(int(p))
                except ValueError:
                    parts.append(0)
            return parts

        cur = parse(current)
        rem = parse(remote)
        max_len = max(len(cur), len(rem))
        cur.extend([0] * (max_len - len(cur)))
        rem.extend([0] * (max_len - len(rem)))
        return rem > cur

    # ── Config ──

    def load_config(self, path: str = OTA_CONFIG_PATH) -> bool:
        if not os.path.isfile(path):
            logger.info(f"OtaUpdater: Config not found at {path}")
            return False
        try:
            with open(path, "r", encoding="utf-8") as f:
                cfg = json.load(f)
            self._manifest_url = cfg.get("manifest_url", "")
            self._auto_check_hours = cfg.get("auto_check_interval_hours", 24)
            self._auto_update = cfg.get("auto_update", False)
            logger.info(f"OtaUpdater: Config loaded, manifest_url={self._manifest_url}")
            return True
        except Exception as e:
            logger.error(f"OtaUpdater: Config load error: {e}")
            return False

    def get_manifest_url(self) -> str:
        return self._manifest_url

    # ── Manifest parsing ──

    def _get_local_version(self, skill_name: str) -> str:
        """Read version from local skill's manifest.json."""
        manifest_path = os.path.join(self._skills_dir, skill_name, "manifest.json")
        if not os.path.isfile(manifest_path):
            return "0.0.0"
        try:
            with open(manifest_path, "r", encoding="utf-8") as f:
                m = json.load(f)
            return m.get("version", "0.0.0")
        except Exception:
            return "0.0.0"

    def parse_manifest(self, manifest_json: str,
                       skills_dir: str = "") -> List[SkillUpdateInfo]:
        """Parse a remote manifest JSON into update info list."""
        if not skills_dir:
            skills_dir = self._skills_dir
        try:
            data = json.loads(manifest_json)
        except Exception:
            return []

        skills = data.get("skills", [])
        updates = []
        for s in skills:
            name = s.get("name", "")
            if not name:
                continue
            remote_ver = s.get("version", "0.0.0")
            local_ver = self._get_local_version(name)
            info = SkillUpdateInfo(
                name=name,
                local_version=local_ver,
                remote_version=remote_ver,
                url=s.get("url", ""),
                sha256=s.get("sha256", ""),
                update_available=self.is_newer_version(local_ver, remote_ver),
                description=s.get("description", ""),
            )
            updates.append(info)
        return updates

    # ── Check for updates ──

    def check_for_updates(self) -> str:
        """Fetch remote manifest and return JSON with available updates."""
        if not self._manifest_url:
            return json.dumps({"error": "No manifest URL configured"})

        try:
            req = urllib.request.Request(self._manifest_url)
            with urllib.request.urlopen(req, context=self._ssl_ctx, timeout=30) as resp:
                manifest_json = resp.read().decode("utf-8")
        except Exception as e:
            return json.dumps({"error": f"Failed to fetch manifest: {e}"})

        updates = self.parse_manifest(manifest_json)
        available = [u for u in updates if u.update_available]

        return json.dumps({
            "available_count": len(available),
            "updates": [
                {
                    "name": u.name,
                    "local_version": u.local_version,
                    "remote_version": u.remote_version,
                    "description": u.description,
                }
                for u in available
            ],
            "total_skills": len(updates),
        })

    # ── Download & Install ──

    def _download_file(self, url: str, dest: str) -> bool:
        try:
            req = urllib.request.Request(url)
            with urllib.request.urlopen(req, context=self._ssl_ctx, timeout=120) as resp:
                with open(dest, "wb") as f:
                    f.write(resp.read())
            return True
        except Exception as e:
            logger.error(f"OtaUpdater: Download failed: {e}")
            return False

    @staticmethod
    def _verify_sha256(filepath: str, expected: str) -> bool:
        if not expected:
            return True  # No checksum to verify
        h = hashlib.sha256()
        with open(filepath, "rb") as f:
            for chunk in iter(lambda: f.read(8192), b""):
                h.update(chunk)
        return h.hexdigest() == expected

    def _backup_skill(self, skill_name: str) -> bool:
        src = os.path.join(self._skills_dir, skill_name)
        if not os.path.isdir(src):
            return True
        os.makedirs(BACKUP_DIR, exist_ok=True)
        dst = os.path.join(BACKUP_DIR, skill_name)
        try:
            if os.path.exists(dst):
                shutil.rmtree(dst)
            shutil.copytree(src, dst)
            return True
        except Exception as e:
            logger.error(f"OtaUpdater: Backup failed for {skill_name}: {e}")
            return False

    def update_skill(self, skill_name: str) -> str:
        """Download and install a specific skill update."""
        if not self._manifest_url:
            return json.dumps({"error": "No manifest URL configured"})

        # Fetch manifest
        try:
            req = urllib.request.Request(self._manifest_url)
            with urllib.request.urlopen(req, context=self._ssl_ctx, timeout=30) as resp:
                manifest_json = resp.read().decode("utf-8")
        except Exception as e:
            return json.dumps({"error": f"Fetch manifest failed: {e}"})

        updates = self.parse_manifest(manifest_json)
        target = None
        for u in updates:
            if u.name == skill_name and u.update_available:
                target = u
                break
        if not target:
            return json.dumps({"error": f"No update available for '{skill_name}'"})

        if not target.url:
            return json.dumps({"error": "No download URL in manifest"})

        # Backup current version
        self._backup_skill(skill_name)

        # Download
        with tempfile.NamedTemporaryFile(suffix=".tar.gz", delete=False) as tmp:
            tmp_path = tmp.name

        try:
            if not self._download_file(target.url, tmp_path):
                return json.dumps({"error": "Download failed"})

            # Verify SHA-256
            if not self._verify_sha256(tmp_path, target.sha256):
                return json.dumps({"error": "SHA-256 verification failed"})

            # Extract
            dest = os.path.join(self._skills_dir, skill_name)
            if os.path.exists(dest):
                shutil.rmtree(dest)
            os.makedirs(dest, exist_ok=True)
            with tarfile.open(tmp_path, "r:gz") as tar:
                tar.extractall(dest)

            # Trigger reload
            if self._reload_cb:
                self._reload_cb()

            return json.dumps({
                "status": "updated",
                "skill": skill_name,
                "from_version": target.local_version,
                "to_version": target.remote_version,
            })
        except Exception as e:
            # Rollback on failure
            self.rollback_skill(skill_name)
            return json.dumps({"error": f"Install failed: {e}"})
        finally:
            if os.path.exists(tmp_path):
                os.unlink(tmp_path)

    # ── Rollback ──

    def rollback_skill(self, skill_name: str) -> str:
        backup = os.path.join(BACKUP_DIR, skill_name)
        if not os.path.isdir(backup):
            return json.dumps({"error": f"No backup found for '{skill_name}'"})

        dest = os.path.join(self._skills_dir, skill_name)
        try:
            if os.path.exists(dest):
                shutil.rmtree(dest)
            shutil.copytree(backup, dest)
            if self._reload_cb:
                self._reload_cb()
            return json.dumps({"status": "rolled_back", "skill": skill_name})
        except Exception as e:
            return json.dumps({"error": f"Rollback failed: {e}"})

    # ── Auto-check loop ──

    async def start_auto_check(self):
        """Start periodic auto-check loop."""
        if not self._manifest_url:
            logger.info("OtaUpdater: No manifest URL, auto-check disabled")
            return
        self._running = True
        self._task = asyncio.create_task(self._auto_check_loop())
        logger.info(f"OtaUpdater: Auto-check started ({self._auto_check_hours}h interval)")

    async def stop(self):
        self._running = False
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass

    async def _auto_check_loop(self):
        while self._running:
            try:
                await asyncio.sleep(self._auto_check_hours * 3600)
                result = self.check_for_updates()
                data = json.loads(result)
                if data.get("available_count", 0) > 0:
                    logger.info(f"OtaUpdater: {data['available_count']} update(s) available")
                    if self._auto_update:
                        for u in data.get("updates", []):
                            self.update_skill(u["name"])
            except asyncio.CancelledError:
                break
            except Exception as e:
                logger.error(f"OtaUpdater: Auto-check error: {e}")

    def get_status(self) -> Dict[str, Any]:
        return {
            "manifest_url": self._manifest_url,
            "auto_check_hours": self._auto_check_hours,
            "auto_update": self._auto_update,
            "running": self._running,
        }

File: src_bak/tizenclaw/infra/ota_updater.cc
/*
 * Copyright (c) 2026 Samsung Electronics Co., Ltd All Rights Reserved
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "ota_updater.hh"

#include <glib.h>

#include <algorithm>
#include <ctime>
#include <filesystem>
#include <fstream>
#include <json.hpp>
#include <sstream>

#include "../common/logging.hh"
#include "http_client.hh"

namespace fs = std::filesystem;

namespace tizenclaw {

OtaUpdater::OtaUpdater(const std::string& skills_dir, ReloadCallback reload_cb)
    : skills_dir_(skills_dir), reload_cb_(std::move(reload_cb)) {
  backup_dir_ = skills_dir_ + "/.backup";
}

bool OtaUpdater::LoadConfig(const std::string& config_path) {
  std::ifstream f(config_path);
  if (!f.is_open()) {
    LOG(WARNING) << "OTA config not found: " << config_path;
    return false;
  }

  try {
    nlohmann::json cfg;
    f >> cfg;

    if (cfg.contains("manifest_url"))
      manifest_url_ = cfg["manifest_url"].get<std::string>();

    if (cfg.contains("auto_check_interval_hours"))
      check_interval_hours_ = cfg["auto_check_interval_hours"].get<int>();

    if (cfg.contains("auto_update"))
      auto_update_ = cfg["auto_update"].get<bool>();

    LOG(INFO) << "OTA config loaded: url=" << manifest_url_;
    return true;
  } catch (const std::exception& e) {
    LOG(ERROR) << "OTA config parse error: " << e.what();
    return false;
  }
}

std::string OtaUpdater::ReadSkillVersion(const std::string& skill_dir) const {
  std::string manifest = skill_dir + "/manifest.json";
  std::ifstream f(manifest);
  if (!f.is_open()) return "0.0.0";

  try {
    nlohmann::json j;
    f >> j;
    if (j.contains("version")) return j["version"].get<std::string>();
  } catch (...) {
  }
  return "0.0.0";
}

bool OtaUpdater::IsNewerVersion(const std::string& local,
                                const std::string& remote) {
  auto parse = [](const std::string& v) -> std::vector<int> {
    std::vector<int> parts;
    std::istringstream iss(v);
    std::string token;
    while (std::getline(iss, token, '.')) {
      try {
        parts.push_back(std::stoi(token));
      } catch (...) {
        parts.push_back(0);
      }
    }
    while (parts.size() < 3) parts.push_back(0);
    return parts;
  };

  auto lv = parse(local);
  auto rv = parse(remote);

  for (size_t i = 0; i < 3; ++i) {
    if (rv[i] > lv[i]) return true;
    if (rv[i] < lv[i]) return false;
  }
  return false;  // Equal
}

std::vector<SkillUpdateInfo> OtaUpdater::ParseManifest(
    const std::string& manifest_json, const std::string& skills_dir) const {
  std::vector<SkillUpdateInfo> updates;

  try {
    auto manifest = nlohmann::json::parse(manifest_json);

    if (!manifest.contains("skills") || !manifest["skills"].is_array())
      return updates;

    for (const auto& skill : manifest["skills"]) {
      SkillUpdateInfo info;
      info.name = skill.value("name", "");
      info.remote_version = skill.value("version", "0.0.0");
      info.download_url = skill.value("url", "");
      info.sha256 = skill.value("sha256", "");

      if (info.name.empty()) continue;

      // Read local version
      std::string skill_dir = skills_dir + "/" + info.name;
      info.local_version = ReadSkillVersion(skill_dir);

      info.update_available =
          IsNewerVersion(info.local_version, info.remote_version);

      updates.push_back(std::move(info));
    }
  } catch (const std::exception& e) {
    LOG(ERROR) << "Manifest parse error: " << e.what();
  }

  return updates;
}

std::string OtaUpdater::CheckForUpdates() {
  if (manifest_url_.empty()) {
    return "{\"error\":"
           "\"No manifest URL configured\"}";
  }

  auto resp = HttpClient::Get(manifest_url_, {}, 2, 10, 30);
  if (!resp.success) {
    return "{\"error\":\"Failed to fetch "
           "manifest: " +
           resp.error + "\"}";
  }

  auto updates = ParseManifest(resp.body, skills_dir_);

  nlohmann::json result;
  result["manifest_url"] = manifest_url_;
  result["updates"] = nlohmann::json::array();

  for (const auto& u : updates) {
    nlohmann::json item = {{"name", u.name},
                           {"local_version", u.local_version},
                           {"remote_version", u.remote_version},
                           {"update_available", u.update_available}};
    result["updates"].push_back(item);
  }

  int available = std::count_if(
      updates.begin(), updates.end(),
      [](const SkillUpdateInfo& u) { return u.update_available; });
  result["available_count"] = available;

  return result.dump();
}

bool OtaUpdater::BackupSkill(const std::string& skill_name) {
  std::string src = skills_dir_ + "/" + skill_name;
  if (!fs::exists(src)) return true;

  // Create backup dir
  std::error_code ec;
  fs::create_directories(backup_dir_, ec);
  if (ec) {
    LOG(ERROR) << "Cannot create backup dir: " << ec.message();
    return false;
  }

  // Backup with timestamp
  auto now = std::time(nullptr);
  std::string backup_name = skill_name + "_" + std::to_string(now);
  std::string dest = backup_dir_ + "/" + backup_name;

  fs::copy(src, dest, fs::copy_options::recursive, ec);
  if (ec) {
    LOG(ERROR) << "Backup failed for " << skill_name << ": " << ec.message();
    return false;
  }

  LOG(INFO) << "Backed up " << skill_name << " to " << dest;
  return true;
}

bool OtaUpdater::DownloadFile(const std::string& url,
                              const std::string& dest_path) {
  auto resp = HttpClient::Get(url, {}, 3, 10, 120);
  if (!resp.success) {
    LOG(ERROR) << "Download failed: " << resp.error;
    return false;
  }

  std::ofstream f(dest_path, std::ios::binary);
  if (!f.is_open()) {
    LOG(ERROR) << "Cannot write to: " << dest_path;
    return false;
  }

  f.write(resp.body.c_str(), static_cast<std::streamsize>(resp.body.size()));
  return true;
}

bool OtaUpdater::VerifySha256(const std::string& file_path,
                              const std::string& expected) const {
  if (expected.empty()) return true;

  std::ifstream f(file_path, std::ios::binary);
  if (!f.is_open()) return false;

  GChecksum* cs = g_checksum_new(G_CHECKSUM_SHA256);
  if (!cs) return false;

  char buf[4096];
  while (f.read(buf, sizeof(buf)) || f.gcount() > 0) {
    g_checksum_update(cs, reinterpret_cast<const guchar*>(buf),
                      static_cast<gssize>(f.gcount()));
  }

  std::string actual = g_checksum_get_string(cs);
  g_checksum_free(cs);

  return actual == expected;
}

std::string OtaUpdater::UpdateSkill(const std::string& skill_name) {
  if (manifest_url_.empty()) {
    return "{\"error\":"
           "\"No manifest URL configured\"}";
  }

  // Fetch manifest
  auto resp = HttpClient::Get(manifest_url_, {}, 2, 10, 30);
  if (!resp.success) {
    return "{\"error\":\"Manifest fetch "
           "failed\"}";
  }

  auto updates = ParseManifest(resp.body, skills_dir_);

  // Find the skill
  const SkillUpdateInfo* target = nullptr;
  for (const auto& u : updates) {
    if (u.name == skill_name) {
      target = &u;
      break;
    }
  }

  if (!target) {
    return "{\"error\":\"Skill not found "
           "in manifest\"}";
  }

  if (!target->update_available) {
    return "{\"status\":\"up_to_date\","
           "\"version\":\"" +
           target->local_version + "\"}";
  }

  // Backup existing skill
  if (!BackupSkill(skill_name)) {
    return "{\"error\":\"Backup failed\"}";
  }

  // Download new version
  std::string tmp_file = "/tmp/tizenclaw_ota_" + skill_name;
  if (!DownloadFile(target->download_url, tmp_file)) {
    return "{\"error\":\"Download failed\"}";
  }

  // Verify checksum
  if (!VerifySha256(tmp_file, target->sha256)) {
    std::error_code ec;
    fs::remove(tmp_file, ec);
    return "{\"error\":\"SHA-256 mismatch\"}";
  }

  // Remove old skill dir & extract new
  std::string skill_dir = skills_dir_ + "/" + skill_name;
  std::error_code ec;
  fs::remove_all(skill_dir, ec);
  fs::create_directories(skill_dir, ec);

  // Move downloaded file as manifest
  // (In production this would be an archive
  //  extraction; for now we treat it as a
  //  direct manifest.json replacement)
  fs::rename(tmp_file, skill_dir + "/manifest.json", ec);
  if (ec) {
    LOG(ERROR) << "Install failed: " << ec.message();
    return "{\"error\":\"Install failed: " + ec.message() + "\"}";
  }

  // Trigger skill reload
  if (reload_cb_) reload_cb_();

  nlohmann::json result = {{"status", "updated"},
                           {"tool", skill_name},
                           {"old_version", target->local_version},
                           {"new_version", target->remote_version}};
  return result.dump();
}

std::string OtaUpdater::RollbackSkill(const std::string& skill_name) {
  // Find latest backup
  std::error_code ec;
  if (!fs::exists(backup_dir_, ec)) {
    return "{\"error\":"
           "\"No backups found\"}";
  }

  std::string latest_backup;
  std::filesystem::file_time_type latest_time{};

  for (const auto& entry : fs::directory_iterator(backup_dir_, ec)) {
    std::string name = entry.path().filename().string();
    if (name.find(skill_name + "_") == 0) {
      auto wt = entry.last_write_time(ec);
      if (!ec && (latest_backup.empty() || wt > latest_time)) {
        latest_backup = entry.path().string();
        latest_time = wt;
      }
    }
  }

  if (latest_backup.empty()) {
    return "{\"error\":\"No backup found "
           "for " +
           skill_name + "\"}";
  }

  // Remove current and restore backup
  std::string skill_dir = skills_dir_ + "/" + skill_name;
  fs::remove_all(skill_dir, ec);
  fs::copy(latest_backup, skill_dir, fs::copy_options::recursive, ec);

  if (ec) {
    return "{\"error\":\"Rollback failed: " + ec.message() + "\"}";
  }

  // Remove used backup
  fs::remove_all(latest_backup, ec);

  // Trigger skill reload
  if (reload_cb_) reload_cb_();

  std::string restored_ver = ReadSkillVersion(skill_dir);

  nlohmann::json result = {{"status", "rolled_back"},
                           {"tool", skill_name},
                           {"restored_version", restored_ver}};
  return result.dump();
}

}  // namespace tizenclaw
